File: mobile-android/app/src/main/java/com/example/p2paudio/protocol/ConnectionCodeCodec.kt
package com.example.p2paudio.protocol

import com.example.p2paudio.model.ConnectionCodePayload

object ConnectionCodeCodec {
    const val PREFIX = "p2paudio-c1:"

    fun looksLikeConnectionCode(raw: String?): Boolean {
        return !raw.isNullOrBlank() && raw.startsWith(PREFIX)
    }

    fun encode(payload: ConnectionCodePayload): String {
        require(payload.host.isNotBlank()) { "Connection code host is blank" }
        require(payload.port in 1..65535) { "Connection code port is out of range" }
        require(payload.token.isNotBlank()) { "Connection code token is blank" }
        require(payload.expiresAtUnixMs > 0L) { "Connection code expiry is invalid" }

        return "$PREFIX${payload.host}:${payload.port}:${payload.expiresAtUnixMs}:${payload.token}"
    }

    fun decode(raw: String): ConnectionCodePayload {
        require(raw.startsWith(PREFIX)) { "Connection code prefix is invalid" }

        val body = raw.removePrefix(PREFIX)
        val parts = body.split(":", limit = 4)
        require(parts.size == 4) { "Connection code format is invalid" }

        val host = parts[0].trim()
        val port = parts[1].trim().toIntOrNull()
        val expiresAtUnixMs = parts[2].trim().toLongOrNull()
        val token = parts[3].trim()

        require(host.isNotBlank()) { "Connection code host is blank" }
        require(port != null && port in 1..65535) { "Connection code port is invalid" }
        require(expiresAtUnixMs != null && expiresAtUnixMs > 0L) { "Connection code expiry is invalid" }
        require(token.isNotBlank()) { "Connection code token is blank" }

        return ConnectionCodePayload(
            host = host,
            port = port,
            token = token,
            expiresAtUnixMs = expiresAtUnixMs
        )
    }
}

File: desktop-windows/src/P2PAudio.Windows.Core/Protocol/ConnectionCodeCodec.cs
using P2PAudio.Windows.Core.Models;

namespace P2PAudio.Windows.Core.Protocol;

public static class ConnectionCodeCodec
{
    public const string Prefix = "p2paudio-c1:";

    public static bool LooksLikeConnectionCode(string? raw)
    {
        return !string.IsNullOrWhiteSpace(raw) &&
               raw.StartsWith(Prefix, StringComparison.Ordinal);
    }

    public static string Encode(ConnectionCodePayload payload)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(payload.Host);
        ArgumentException.ThrowIfNullOrWhiteSpace(payload.Token);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(payload.Port);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(payload.ExpiresAtUnixMs);

        return $"{Prefix}{payload.Host}:{payload.Port}:{payload.ExpiresAtUnixMs}:{payload.Token}";
    }

    public static ConnectionCodePayload Decode(string raw)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(raw);
        if (!raw.StartsWith(Prefix, StringComparison.Ordinal))
        {
            throw new ArgumentException("Connection code prefix is invalid.", nameof(raw));
        }

        var body = raw[Prefix.Length..];
        var parts = body.Split(':', 4, StringSplitOptions.None);
        if (parts.Length != 4)
        {
            throw new ArgumentException("Connection code format is invalid.", nameof(raw));
        }

        var host = parts[0].Trim();
        var portRaw = parts[1].Trim();
        var expiresAtRaw = parts[2].Trim();
        var token = parts[3].Trim();

        if (string.IsNullOrWhiteSpace(host) ||
            !int.TryParse(portRaw, out var port) ||
            port <= 0 ||
            port > ushort.MaxValue ||
            !long.TryParse(expiresAtRaw, out var expiresAtUnixMs) ||
            expiresAtUnixMs <= 0 ||
            string.IsNullOrWhiteSpace(token))
        {
            throw new ArgumentException("Connection code payload is invalid.", nameof(raw));
        }

        return new ConnectionCodePayload(
            Host: host,
            Port: port,
            Token: token,
            ExpiresAtUnixMs: expiresAtUnixMs
        );
    }
}
